src/gui/AutomatableModelView.cpp

```diff
@@ -258,7 +279,228 @@ void AutomatableModelViewSlots::removeConnection()
 }
 
 
+void AutomatableModelViewSlots::addSongAutomationNode()
+{
+	// getting all the clips that have this model
+	std::vector<AutomationClip*> clips = AutomationClip::clipsForModel(m_amv->modelUntyped());
+	// selecting the track with the most clips connected to this model
+	AutomationTrack* track = getCurrentAutomationTrack(&clips, true);
+	// getting the clip before the current song time position
+	AutomationClip* clip = getCurrentAutomationClip(track, true, false);
+
+	// getting global song time
+	TimePos timePos = static_cast<TimePos>(Engine::getSong()->getPlayPos());
+	// account for the node's relative position inside clip
+	timePos -= clip->startPosition();
+	bool autoResize = clip->getAutoResize();
+
+	clip->setAutoResize(true);
+	// adding model value
+	clip->recordValue(timePos, m_amv->modelUntyped()->getTrueValue());
+	clip->setAutoResize(autoResize);
+}
+void AutomatableModelViewSlots::addSongAutomationNodeAndClip()
+{
+	std::vector<AutomationClip*> clips = AutomationClip::clipsForModel(m_amv->modelUntyped());
+	AutomationTrack* track = getCurrentAutomationTrack(&clips, true);
+	AutomationClip* clip = getCurrentAutomationClip(track, false, false);
+
+	TimePos timePos = static_cast<TimePos>(Engine::getSong()->getPlayPos());
 
+	if (clip != nullptr && clip->endPosition().getTicks() < timePos.getTicks())
+	{
+		AutomationClip* newClip = makeNewClip(track, timePos, true);
+		// copying the progressionType of the clip before
+		newClip->setProgressionType(clip->progressionType());
+		timePos -= newClip->startPosition();
+		bool autoResize = newClip->getAutoResize();
+
+		newClip->setAutoResize(true);
+		newClip->recordValue(timePos, m_amv->modelUntyped()->getTrueValue());
+		newClip->setAutoResize(autoResize);
+	}
+	else
+	{
+		addSongAutomationNode();
+	}
+}
+void AutomatableModelViewSlots::updateSongNearestAutomationNode()
+{
+	std::vector<AutomationClip*> clips = AutomationClip::clipsForModel(m_amv->modelUntyped());
+	// getting the track without adding a new one if no track was found
+	AutomationTrack* track = getCurrentAutomationTrack(&clips, false);
+	// this needs to be checked because getCurrentAutomationTrack might give
+	// a nullptr if it can not find and add a track
+	if (track != nullptr)
+	{
+		// getting nearest node position
+		AutomationClip* nodeClip = nullptr;
+		TimePos nodePos = getNearestAutomationNode(track, &nodeClip);
+		if (nodeClip != nullptr)
+		{
+			// modifying its value
+			nodeClip->recordValue(nodePos, m_amv->modelUntyped()->getTrueValue());
+		}
+	}
+}
+void AutomatableModelViewSlots::removeSongNearestAutomationNode()
+{
+	std::vector<AutomationClip*> clips = AutomationClip::clipsForModel(m_amv->modelUntyped());
+	// getting the track without adding a new one if no track was found
+	AutomationTrack* track = getCurrentAutomationTrack(&clips, false);
+	// this needs to be checked because getCurrentAutomationTrack might give
+	// a nullptr if it can not find and add a track
+	if (track != nullptr)
```

Comment:
You can invert here too for early return

Reply:
What do you mean by invert?

Reply:
https://softwareengineering.stackexchange.com/questions/18454/should-i-return-from-a-function-early-or-use-an-if-statement

Might help

Reply:
Or if you don't understand that one, https://medium.com/swlh/return-early-pattern-3d18a41bba8

Basically the same thing but with more context

Reply:
I generally try to avoid multiple returns to make the code more readable.
I think adding returns for example in the first lines of functions is fine, because they will not be hidden inside the code.
In this case there could be 2 returns added, both after the start and before the end. I think it makes that function mode unreadable.

Do you still want to change this?

Reply:
Just return the one at the start. The idea is to reduce nesting to a reasonable amount